velox/functions/lib/Re2Functions.cpp

```diff
@@ -666,6 +775,97 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
       rows);
 }
 
+// Match the input against a kRelaxedSubString pattern.
+// The algorithm is:
+// - Start by searching the first literal sub-pattern in the input.
+// - If we find the sub-pattern, we match the left & right part of the whole
+// pattern.
+// - If we failed matching any sub-pattern, we continue by searching the failed
+// literal sub-pattern.
+template <bool isAscii>
+bool matchRelaxedSubstringPattern(
+    const StringView& input,
+    const PatternMetadata& patternMetadata) {
+  if (input.size() < patternMetadata.length()) {
+    return false;
+  }
+
+  // Search the specified literal pattern in the input.
+  auto inputView = std::string_view(input);
+  auto patternStrHeader = patternMetadata.fixedPattern().c_str();
+  // The index in the input we are matching against.
+  auto currentMatchingOffsetInInput = 0;
+  auto searchLiteralPattern = [&](SubPatternMetadata& subPattern) -> int32_t {
+    auto index = inputView.find(
+        patternStrHeader + subPattern.start,
+        currentMatchingOffsetInInput,
+        subPattern.length);
+    if (index == std::string::npos) {
+      return -1;
+    }
+
+    return index;
+  };
+
+  // The index of the sub literal pattern we are matching.
+  auto currentPatternIndex =
+      (patternMetadata.subPatterns()[0].kind == kLiteralString ? 0 : 1);
+  auto currentSubPattern = patternMetadata.subPatterns()[currentPatternIndex];
+  auto end = input.size();
+  while (currentMatchingOffsetInInput < end) {
+    // Search current literal pattern to find out where to start matching
+    // the whole pattern.
+    currentMatchingOffsetInInput = searchLiteralPattern(currentSubPattern);
+    // If any literal pattern is not found, fails the matching.
+    if (currentMatchingOffsetInInput < 0) {
+      return false;
+    }
+
+    std::pair<bool, int32_t> matchResult;
+    // Match the right part of the pattern(if there is a right part).
+    if (currentPatternIndex < patternMetadata.numSubPatterns() - 1) {
+      matchResult = matchRelaxedFixedForward<isAscii>(
+          input,
+          patternMetadata,
+          currentMatchingOffsetInInput + currentSubPattern.length,
+          currentPatternIndex + 1);
+    }
+
+    // Match the left part of the pattern(if there is a left part).
+    if ((currentPatternIndex == patternMetadata.numSubPatterns() - 1) ||
```

Comment:
I believe this code works, but I find it hard to follow. I wonder if there is a way to make it a bit more readable. Some ideas that come to mind:

1- replace `std::pair<bool, int32_t>` with a specific struct that uses intuitive names for fields and provides helper methods like isFullMatch(), isPartialMatch(), etc.

1- change the flow of this code to (1) match forward, check results and exit early; (2) match backward, check results.

... match forward

if (matchResult.cannotMatch()) {
   return false;
}

if (matchResult.partialMatch()) {
 ...
 continue;
}

....match backwards

if (matchResult.cannotMatch()) {
   return false;
}

if (matchResult.partialMatch()) {
 ...
 continue;
}